test/common/upstream/hds_test.cc

```diff
@@ -480,6 +486,86 @@ TEST_F(HdsTest, TestMinimalOnReceiveMessage) {
   hds_delegate_->onReceiveMessage(std::move(message));
 }
 
+// Test that a transport_socket_matches and transport_socket_match_criteria filter as expected to
+// build the correct TransportSocketFactory based on these fields.
+TEST_F(HdsTest, TestSocketContext) {
+  EXPECT_CALL(*async_client_, startRaw(_, _, _, _)).WillOnce(Return(&async_stream_));
+  EXPECT_CALL(async_stream_, sendMessageRaw_(_, _));
+  createHdsDelegate();
+
+  // Create Message.
+  message.reset(createSimpleMessage());
+
+  // Add transport socket matches to message.
+  const std::string match_yaml = absl::StrFormat(
+      R"EOF(
+name: "test_socket"
+match:
+  test_match: "true"
+transport_socket:
+  name: "envoy.transport_sockets.raw_buffer"
+ )EOF");
+  auto* cluster_health_check = message->mutable_cluster_health_checks(0);
```

Comment:
You should be able to build up the entire cluster health check in YAML and then interpret via `loadFromYaml`.

Reply:
The only issue with this is since I'm pulling `cluster_health_check` from createSimpleMessage(), it is not possible to do an all-in-one merge with cluster_health_check because when merging `cluster_health_check.health_checks()` it is a repeated field so it appends to the list instead of updating health_check(0), which is what I need to be doing in this case. So I have to build two different in-line yaml samples, one for `cluster_health_check` and one for its child `cluster_health_check.health_checks()`, and merge from there. 

This definitely reduced the complexity of trying to build the `transport_socket_match_criteria` from scratch.